### ISSUE TITLE
Add new version to already created revision

### ISSUE THREAD

**Issue by platico** [etianen/django-reversion#672]
I have created a "backup" functionality using the Django Reversion library for an object that keeps revision of a registered model, including the related objects (foreign keys etc.).

Let's use the following case as an example.

```
class Newspaper:
   title = models.CharField()

class Article:
   newspaper = models.ForeignKey('Newspaper') 
```

When the user clicks to keep a backup of a newspaper object, a revision of the newspaper is created along with the already created articles, under the same revision. I did it so that when the user chooses to revert back to the latest backup, all related objects under the same revision to be reverted.

The problem starts when a new article of the newspaper is created after the revision is created. The issue is that if the user chooses to revert to the previous revision (the one before creating the new article), the new article will still be there because it was not registered in the latest revision.

Furthermore, I don't want to create a new revision every time a new article is created because then there might be other changes included that the user will not want to include to the revision.

What I think might be a suitable solution, is when a new article is created, to create a Version of the article and register it in the latest created revision of the newspaper, but I don't know if it's possible using the functions provided by the API (maybe using add_to_revision?)

**Comment by etianen** [comment]
I take it you're reverting using: .revision.revert()

Try this: .revision.revert(delete=True)

:)

**Comment by platico** [comment]
Thanks for the quick answer! I already use revert(delete=True) but for some reason the newly created objects are not versioned. Actually my case is a little more complicated and now I am worried that my objects' registration is incorrect. The actual structure is like this:


```
class Newspaper:
   title = models.CharField()

class Article:
   newspaper = models.ForeignKey('Newspaper')

class Paragraph:
   article = models.ForeignKey('Article')  
```

The way I register the Newspaper revision is when a user clicks to create a backup, the `register.create_revision()` block is called and then within this block I save the newspaper and the related objects manually.


```
with reversion.create_revision():
       newspaper.save(create_revision=True)         
```     


and then in the .save(create_revision=False, user=None) function of Newspaper:

```
def save(create_revision=False, user=None):
       if create_revision:
             reversion.set_user(user)
             reversion.set_comment("Created revision for Newsletter Config")
             for article in self.articles.all():
                    article.save(create_revision=True)
```

and the process continues recursively for paragraphs..

I also have set the follow variable when registering the models in the admin.py, so for Newspaper it would be:

```
@admin.register(Newspaper)
class NewpaperAdmin(VersionAdmin):

    pass

@admin.register(Article, follow=['newspaper'])
class ArticleAdmin(VersionAdmin):

    pass

```

**Comment by etianen** [comment]
You are not registering your models correctly. Try this:

@admin.register(Newspaper)
class NewpaperAdmin(VersionAdmin):

    def reversion_register(self, model, **options):
        options["follow"] = ("article_set",)
        super(NewpaperAdmin, self).reversion_register(model, **options)

@admin.register(Article])
class ArticleAdmin(VersionAdmin):

    def reversion_register(self, model, **options):
        options["follow"] = ("newspaper",)
        super(ArticleAdmin, self).reversion_register(model, **options)

**Comment by platico** [comment]
Changed the admin.py file using the structure you suggested but now I get the following error:

`NameError: global name 'NewspaperAdmin' is not defined
`

Have you seen it before?

**Comment by etianen** [comment]
Looks like a typo. I guess in my code, NewpaperAdmin should become New*s*
paperAdmin

**Comment by platico** [comment]
Ok fixed. One last thing as the revert function I call it like:

`Version.objects.get_for_object(newspaper_object_to_revert)[0].revision.revert(delete=True)`

Is the usage correct? Thanks

**Comment by etianen** [comment]
That'll do fine, I think.

**Comment by platico** [comment]
Ok registration works now thank you! My initial problem persists though. The problem is that when I revert(), the related objects that are in the same revision are reverted, but the ones that were created after the latest revision was created are not, as they remain unregistered.

The (delete=True) argument deletes the objects that are no longer in the current revision but not the ones that are created after the latest revision.

**Comment by etianen** [comment]
Which related objects are not being deleted?

**Comment by platico** [comment]
I will draw it to make it more clear. Lets assume I have created a newspaper object(Newspaper0), an article object(Article0) that has a foreign key to the newspaper and a paragraph object(Paragraph0) that has a foreign key to the article. At this point I create a backup, hence a revision (revision0) of the newspaper is created that includes a version for the 3 objects. After the backup is created, I create a 2nd paragraph that also points to the article object(Article0). But, since it was created after the creation of revision0, it is not included in revision0. 

Now, if I revert to revision0, all the applied changes of Newspaper0, Article0 and Paragraph0 will revert to their previous status, but Paragraph1 will not be deleted, even though it hadn't been created when revision0 was created.

```
Newspaper0
           |
           |----Article0
                     |
                     |-------Paragraph0
====================================================Backup Point (revision0)
                     |-------Paragraph1
====================================================Current Point
```

**Comment by etianen** [comment]
Then try this registration code:

@admin.register(Newspaper)
class NewspaperAdmin(VersionAdmin):

    def reversion_register(self, model, **options):
        options["follow"] = ("article_set",)
        super(NewspaperAdmin, self).reversion_register(model, **options)

@admin.register(Article])
class ArticleAdmin(VersionAdmin):

    def reversion_register(self, model, **options):
        options["follow"] = ("newspaper", *"paragraph_set",*)
        super(ArticleAdmin, self).reversion_register(model, **options)


You must include all relations you consider part of the same revision in
the "follow" argument to reversion.register().

**Comment by platico** [comment]
That means that when a new object is created (ex. paragraph), it is automatically registered in the latest revision of the object(s) it follows?

**Comment by etianen** [comment]
The *follow *argument tells reversion which objects belong in the same
logical revision. When any object is saved, the *follow* argument will be
used to traverse database relations to save a copy of the related object in
the revision.

The *follow *argument is also used in revert(delete=True) to determine what
relations should be deleted during the revert.

**Comment by platico** [comment]
Ok I think I get it now! That subsequently means that when I call reversion.create_revision() for a newspaper object (and I within it I call newspaper_object.save()), do I also have to call .save() for the followed objects or is it done automatically?

**Comment by etianen** [comment]
The *follow* argument ensures .save() is called on all followed relations
automatically if any of the related models are saved. (Not quite true, it
doesn't actually call save(), it just adds them to the revision, but it's
the same from reversion's point of view).

Basically, use the *follow* argument to define all relations that belong in
the same revision, and everything should just work. The problem you had was
that *paragraph_set* wasn't included in the *follow* configuration, so
paragraphs weren't deleted during revert(delete=True).

**Comment by platico** [comment]
Ok for now is fixed but in order to fix it, I changed the way it is registered, which creates errors.

The specific use case is that in admin.py I declare the NewspaperAdmin like this:

```
@admin.register(Newspaper)
class NewspaperAdmin(VersionAdmin):

     def reversion_register(self, model, **options):
        options["follow"] = ("article_set",)
        super(NewspaperAdmin, self).reversion_register(model, **options)

@admin.register(Article)
class ArticleAdmin(VersionAdmin):

    def reversion_register(self, model, **options):
        options["follow"] = ("newspaper",)
        super(ArticleAdmin, self).reversion_register(model, **options)

```

In this case I get `NameError: global name 'NewspaperAdmin' is not defined` error.

On the other hand, when I change to:

```
@admin.register(Newspaper)
class NewspaperAdmin(VersionAdmin):

     pass
```

...then the error is gone, but unfortunately in this case I cannot use follow...

The way I trespassed this error was by not using `@admin.register(Newspaper)` but by using `admin.site.register(Newspaper, NewspaperAdmin)` but this case produces errors with parent classes.

To sum up, I think the problem is between `@admin.register()` decorator and the `reversion_register()` function. Have you ever had similar cases? thanks

**Comment by etianen** [comment]
Can you provide a full traceback?

In general, a NameError means you typo'd the name of a class or variable.

**Comment by platico** [comment]
The traceback I get is this:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/core/management/__init__.py", line 351, in execute_from_command_line
    utility.execute()
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/core/management/__init__.py", line 325, in execute
    django.setup()
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/contrib/admin/apps.py", line 22, in ready
    self.module.autodiscover()
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/contrib/admin/__init__.py", line 24, in autodiscover
    autodiscover_modules('admin', register_to=site)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/utils/module_loading.py", line 74, in autodiscover_modules
    import_module('%s.%s' % (app_config.name, module_to_search))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/platico/PycharmProjects/news-service/src/configs/admin.py", line 19, in <module>
    class NewspaperAdmin(VersionAdmin):
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/contrib/admin/decorators.py", line 25, in _model_admin_wrapper
    admin_site.register(models, admin_class=admin_class)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 108, in register
    self._registry[model] = admin_class(model, self)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/reversion/admin.py", line 152, in __init__
    self._reversion_autoregister(self.model, inline_fields)
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/reversion/admin.py", line 104, in _reversion_autoregister
    self._reversion_autoregister(parent_model, ())
  File "/Users/platico/PycharmProjects/news-service/src/lib/python2.7/site-packages/reversion/admin.py", line 105, in _reversion_autoregister
    self.reversion_register(model, follow=follow)
  File "/Users/platico/PycharmProjects/news-service/src/configs/admin.py", line 22, in reversion_register
    super(NewspaperAdmin, self).reversion_register(model, **kwargs)
NameError: global name 'NewspaperAdmin' is not defined
```

Also Newspaper inherits from TimestampModel and PeriodicTask classes. As a result, the Newspaper class looks like this:

```
class Newspaper(PeriodicTask, TimeStampedModel):
```

I don't know if it plays a role but just to mention...

**Comment by platico** [comment]
Also I get the same NameError with other classes that inherit from `models.Model`, so I don't think that PeriodicTask or TimestampedModel play a role on the error

**Comment by etianen** [comment]
And you definitely have a class called NewspaperAdmin?

If you are on Python 3, try using:

super().reversion_register(model, **kwargs)

Instead of:

super(NewspaperAdmin, self).reversion_register(model, **kwargs)

**Comment by platico** [comment]
I somehow managed to solve it by **not** using the decorator `@admin.register(Newspaper)` but by registering the class in the admin using `admin.site.register(Newspaper, NewspaperAdmin)`. Then the error disappears.
